monkey/monkey_island/cc/ui/src/components/ui-components/plugins-marketplace/AvailablePlugins.tsx

```diff
@@ -0,0 +1,133 @@
+import React, {useContext, useState} from 'react';
+import {installPlugin} from './mocksHelper';
+import {shallowAdditionOfUniqueValueToArray, shallowRemovalOfUniqueValueFromArray} from '../../../utils/objectUtils';
+import {PluginsContext} from './PluginsContext';
+import {GridActionsCellItem} from '@mui/x-data-grid';
+import {nanoid} from 'nanoid';
+import FileDownloadIcon from '@mui/icons-material/FileDownload';
+import DownloadDoneIcon from '@mui/icons-material/DownloadDone';
+import DownloadingIcon from '@mui/icons-material/Downloading';
+import BasePlugins from './BasePlugins';
+import SearchBar from '../SearchBar';
+import {Box} from '@mui/material';
+import AuthComponent from '../../AuthComponent';
+
+// Provides the plugins, filtering out the installed plugins
+class AvailablePluginsView {
+  availablePlugins = []; // all plugins, grouped by type then name
+  installedPlugins = []; // all installed plugins, grouped by type then name
+
+  constructor(availablePlugins, installedPlugins) {
+    this.availablePlugins = availablePlugins;
+    this.installedPlugins = installedPlugins;
+  }
+
+  pluginInstalled(plugin_type, plugin_name) {
+    return plugin_type in this.installedPlugins && plugin_name in this.installedPlugins[plugin_type];
+  }
+
+  * makeAvailablePluginsIterator() {
+    for (const plugin_type in this.availablePlugins) {
+      for (const plugin_name in this.availablePlugins[plugin_type]) {
+        if (!this.pluginInstalled(plugin_type, plugin_name)) {
+          // There may be multiple versions of the same plugin, so we only want the latest one
+          yield this.availablePlugins[plugin_type][plugin_name].slice(-1)[0];
+        }
+      }
+    }
+  }
+
+  [Symbol.iterator]() {
+    return this.makeAvailablePluginsIterator();
+  }
```

Comment:
This is object-oriented approach in an otherwise functional code. We can transform the data if we need a list, but creating our own collection that's sometimes object sometimes list complicates usage